<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Benjamin Leblanc" />


<title>Genomic regions for BRD normalization</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Genomic regions for BRD normalization</h1>
<h3 class="subtitle"><em>Tightrope version 0.7.1 - 02.06.2018</em></h3>
<h4 class="author"><em><a href="https://github.com/benja0x40">Benjamin Leblanc</a></em></h4>


<div id="TOC">
<ul>
<li><a href="#summary"><span class="toc-section-number">1</span> Summary</a></li>
<li><a href="#genomic-annotations-from-ensembl-and-ucsc"><span class="toc-section-number">2</span> Genomic annotations from Ensembl and UCSC</a><ul>
<li><a href="#provided-genomic-regions-in-human-and-mouse"><span class="toc-section-number">2.1</span> Provided genomic regions in human and mouse</a></li>
<li><a href="#making-genomic-regions-for-other-organims-or-genome-assemblies"><span class="toc-section-number">2.2</span> Making genomic regions for other organims or genome assemblies</a></li>
</ul></li>
<li><a href="#genomic-tiling-with-regular-bins"><span class="toc-section-number">3</span> Genomic tiling with regular bins</a></li>
<li><a href="#defining-other-relevant-regions"><span class="toc-section-number">4</span> Defining other relevant regions</a></li>
</ul>
</div>

<hr />
<p>The purpose of this documentation is to provide a practical guide to the use of <code>R</code> functions available in <code>Tightrope</code>. A basic knowledge of the <code>R</code> syntax and data structures, and of <a href="http://www.bioconductor.org/">Bioconductor</a> objects used to represent and manipulate mapped reads and genomic intervals will help to understand and adapt the included source code examples to other ChIP-seq datasets.</p>
<hr />

<div id="summary" class="section level1">
<h1><span class="header-section-number">1</span> Summary</h1>
</div>
<div id="genomic-annotations-from-ensembl-and-ucsc" class="section level1">
<h1><span class="header-section-number">2</span> Genomic annotations from Ensembl and UCSC</h1>
<div id="provided-genomic-regions-in-human-and-mouse" class="section level2">
<h2><span class="header-section-number">2.1</span> Provided genomic regions in human and mouse</h2>
<div id="overview" class="section level3">
<h3><span class="header-section-number">2.1.1</span> Overview</h3>
<p>To simplify the generation of read count matrixes, <code>Tightrope</code> provides genomic regions in human and mouse which correspond to annotated gene features from the Ensembl database, as well as CpG-islands from UCSC and blacklisted regions from the ENCODE project.</p>
</div>
<div id="usage" class="section level3">
<h3><span class="header-section-number">2.1.2</span> Usage</h3>
<p>The <code>summary</code> element in <code>EGA91_human</code> and <code>EGA91_mouse</code> objects contains a brief description of provided gene features.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(Tightrope)  <span class="co"># Load package</span>
<span class="kw">data</span>(<span class="st">&quot;EGA91_human&quot;</span>) <span class="co"># Load R data with gene features in Human</span>
EGA91_human<span class="op">$</span>summary <span class="co"># Summary of loaded gene features</span></code></pre></div>
<p>For instance, gene units in human according to Ensembl GRCh38.p10 release 91 are provided by the <code>GRanges</code> object <code>EGA91_human$GNU</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">EGA91_human<span class="op">$</span>GNU <span class="co"># Gene units</span></code></pre></div>
<pre><code>## GRanges object with 58243 ranges and 1 metadata column:
##                   seqnames              ranges strand |   gene_name
##                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   ENSG00000000003     chrX 100627109-100639991      - |      TSPAN6
##   ENSG00000000005     chrX 100584802-100599885      + |        TNMD
##   ENSG00000000419    chr20   50934867-50958555      - |        DPM1
##   ENSG00000000457     chr1 169849631-169894267      - |       SCYL3
##   ENSG00000000460     chr1 169662007-169854080      + |    C1orf112
##               ...      ...                 ...    ... .         ...
##   ENSG00000284744     chr1     6767954-6770038      + |            
##   ENSG00000284745     chr1     2960658-2968707      - |            
##   ENSG00000284746     chr8   12601158-12601376      - |            
##   ENSG00000284747     chr1     7991134-8005312      + |            
##   ENSG00000284748     chr1   37596126-37607336      + |            
##   -------
##   seqinfo: 25 sequences (1 circular) from hg38 genome</code></pre>
</div>
<div id="human-genome" class="section level3">
<h3><span class="header-section-number">2.1.3</span> Human genome</h3>
<div id="grch38.p10-hg38" class="section level4">
<h4><span class="header-section-number">2.1.3.1</span> GRCh38.p10 / hg38</h4>
<center>
<strong>Genomic regions</strong>
</center>
<table>
<thead>
<tr class="header">
<th>R data object</th>
<th>Organism</th>
<th>Genome build</th>
<th>Source</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>EGA91_human</td>
<td>Homo sapiens</td>
<td>GRCh38 (hg38)</td>
<td>Ensembl genes (rel. 91, Dec. 2017)</td>
</tr>
<tr class="even">
<td>CGI_hg38</td>
<td>Homo sapiens</td>
<td>GRCh38 (hg38)</td>
<td>UCSC / cpgIslandExt.txt.gz</td>
</tr>
<tr class="odd">
<td>hg38_blacklist</td>
<td>Homo sapiens</td>
<td>GRCh38 (hg38)</td>
<td>ENCODE</td>
</tr>
</tbody>
</table>
<p> </p>
<table>
<caption>Summary of gene features in EGA91_human</caption>
<thead>
<tr class="header">
<th align="left">object</th>
<th align="left">description</th>
<th align="right">size</th>
<th align="right">coverage</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">INTERGENIC</td>
<td align="left">intergenic regions</td>
<td align="right">33461</td>
<td align="right">1478323127</td>
</tr>
<tr class="even">
<td align="left">GNU</td>
<td align="left">gene units</td>
<td align="right">58243</td>
<td align="right">1609963274</td>
</tr>
<tr class="odd">
<td align="left">TXU</td>
<td align="left">transcribed units</td>
<td align="right">200243</td>
<td align="right">1609953929</td>
</tr>
<tr class="even">
<td align="left">TSS</td>
<td align="left">transcription start sites</td>
<td align="right">200243</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">TTS</td>
<td align="left">transcription termination sites</td>
<td align="right">200243</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">CCS</td>
<td align="left">constitutive coding sequences (i.e. ‘purely’ exon)</td>
<td align="right">118829</td>
<td align="right">19916386</td>
</tr>
<tr class="odd">
<td align="left">CNC</td>
<td align="left">constitutive intragenic non-coding sequences</td>
<td align="right">242020</td>
<td align="right">1574360321</td>
</tr>
<tr class="even">
<td align="left">ACS</td>
<td align="left">alternative coding sequence (not CCS or CNC)</td>
<td align="right">111118</td>
<td align="right">15686567</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Human genome annotations (Ensembl GRCh38.p10 release 91)</span>
<span class="kw">data</span>(<span class="st">&quot;EGA91_human&quot;</span>)    <span class="co"># Gene features from Ensembl      (environment)</span>
<span class="kw">data</span>(<span class="st">&quot;CGI_hg38&quot;</span>)       <span class="co"># CpG Islands from UCSC           (GRanges)</span>
<span class="kw">data</span>(<span class="st">&quot;hg38_blacklist&quot;</span>) <span class="co"># Blacklisted regions from ENCODE (GRanges)</span></code></pre></div>
</div>
</div>
<div id="mouse-genome" class="section level3">
<h3><span class="header-section-number">2.1.4</span> Mouse genome</h3>
<div id="grcm38.p5-mm10" class="section level4">
<h4><span class="header-section-number">2.1.4.1</span> GRCm38.p5 / mm10</h4>
<center>
<strong>Genomic regions</strong>
</center>
<table>
<thead>
<tr class="header">
<th>R data object</th>
<th>Organism</th>
<th>Genome build</th>
<th>Source</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>EGA91_mouse</td>
<td>Mus musculus</td>
<td>GRCm38 (mm10)</td>
<td>Ensembl genes (rel. 91, Dec. 2017)</td>
</tr>
<tr class="even">
<td>CGI_mm10</td>
<td>Mus musculus</td>
<td>GRCm38 (mm10)</td>
<td>UCSC / cpgIslandExt.txt.gz</td>
</tr>
<tr class="odd">
<td>mm10_blacklist</td>
<td>Mus musculus</td>
<td>GRCm38 (mm10)</td>
<td>ENCODE</td>
</tr>
</tbody>
</table>
<p> </p>
<table>
<caption>Summary of gene features in EGA91_mouse</caption>
<thead>
<tr class="header">
<th align="left">object</th>
<th align="left">description</th>
<th align="right">size</th>
<th align="right">coverage</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">INTERGENIC</td>
<td align="left">intergenic regions</td>
<td align="right">36878</td>
<td align="right">1527434550</td>
</tr>
<tr class="even">
<td align="left">GNU</td>
<td align="left">gene units</td>
<td align="right">53379</td>
<td align="right">1198103119</td>
</tr>
<tr class="odd">
<td align="left">TXU</td>
<td align="left">transcribed units</td>
<td align="right">133849</td>
<td align="right">1195980453</td>
</tr>
<tr class="even">
<td align="left">TSS</td>
<td align="left">transcription start sites</td>
<td align="right">133849</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">TTS</td>
<td align="left">transcription termination sites</td>
<td align="right">133849</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">CCS</td>
<td align="left">constitutive coding sequences (i.e. ‘purely’ exon)</td>
<td align="right">153477</td>
<td align="right">26801776</td>
</tr>
<tr class="odd">
<td align="left">CNC</td>
<td align="left">constitutive intragenic non-coding sequences</td>
<td align="right">241380</td>
<td align="right">1161593764</td>
</tr>
<tr class="even">
<td align="left">ACS</td>
<td align="left">alternative coding sequence (not CCS or CNC)</td>
<td align="right">69520</td>
<td align="right">9707579</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Mouse genome annotations (Ensembl GRCm38.p5 release 91)</span>
<span class="kw">data</span>(<span class="st">&quot;EGA91_mouse&quot;</span>)    <span class="co"># Gene features from Ensembl      (environment)</span>
<span class="kw">data</span>(<span class="st">&quot;CGI_mm10&quot;</span>)       <span class="co"># CpG Islands from UCSC           (GRanges)</span>
<span class="kw">data</span>(<span class="st">&quot;mm10_blacklist&quot;</span>) <span class="co"># Blacklisted regions from ENCODE (GRanges)</span></code></pre></div>
</div>
</div>
</div>
<div id="making-genomic-regions-for-other-organims-or-genome-assemblies" class="section level2">
<h2><span class="header-section-number">2.2</span> Making genomic regions for other organims or genome assemblies</h2>
<div id="overview-1" class="section level3">
<h3><span class="header-section-number">2.2.1</span> Overview</h3>
<p>The 4 next examples show how to generate R objects with gene annotations for the human and mouse genomes, adding the following genome releases to the ones presented in the previous section.</p>
<p>In order to save R objects transparently, the next examples use the LittleThumb package which can be installed as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;devtools&quot;</span>) <span class="co"># (devtools can be installed from CRAN repositories)</span>
<span class="kw">install_github</span>(<span class="st">&quot;benja0x40/LittleThumb&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(LittleThumb)</code></pre></div>
</div>
<div id="human-genome-1" class="section level3">
<h3><span class="header-section-number">2.2.2</span> Human genome</h3>
<div id="grch38.p10-hg38-1" class="section level4">
<h4><span class="header-section-number">2.2.2.1</span> GRCh38.p10 / hg38</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># =============================================================================.</span>
<span class="co"># Human genome annotations (Ensembl GRCh38.p10 release 91)</span>
<span class="co"># -----------------------------------------------------------------------------.</span>
organism &lt;-<span class="st"> </span><span class="kw">list</span>(
  <span class="dt">name    =</span> <span class="st">&quot;Homo sapiens&quot;</span>,
  <span class="dt">taxid   =</span> <span class="dv">9606</span>,
  <span class="dt">UCSC    =</span> <span class="kw">list</span>(<span class="dt">genome =</span> <span class="st">&quot;hg38&quot;</span>),
  <span class="dt">Ensembl =</span> <span class="kw">list</span>(<span class="dt">release =</span> <span class="dv">91</span>),
  <span class="dt">BSg_pkg =</span> <span class="st">&quot;BSgenome.Hsapiens.UCSC.hg38&quot;</span>,
  <span class="dt">TxDb =</span> <span class="kw">list</span>(
    <span class="dt">source =</span> <span class="st">&quot;Ensembl GRCh38.p10 release 91 (December 2017)&quot;</span>,
    <span class="dt">portal =</span> <span class="st">&quot;http://Dec2017.archive.ensembl.org/Homo_sapiens&quot;</span>,
    <span class="dt">gtf =</span> <span class="kw">paste0</span>(
      <span class="st">&quot;ftp://ftp.ensembl.org/pub/&quot;</span>,
      <span class="st">&quot;release-91/gtf/homo_sapiens/Homo_sapiens.GRCh38.91.gtf.gz&quot;</span>
    )
  ),
  <span class="dt">BioMart =</span> <span class="kw">list</span>(
    <span class="dt">host     =</span> <span class="st">&quot;www.ensembl.org&quot;</span>, <span class="co"># Dec2017.archive.ensembl.org</span>
    <span class="dt">database =</span> <span class="st">&quot;ensembl&quot;</span>,
    <span class="dt">dataset  =</span> <span class="st">&quot;hsapiens_gene_ensembl&quot;</span>
  )
)

<span class="co"># Gene features from Ensembl</span>
<span class="kw">MakeObj</span>(EGA91_human, {
  EGA91_human &lt;-<span class="st"> </span><span class="kw">BuildGeneFeatures</span>(organism)
})

<span class="co"># CpG Islands from UCSC</span>
<span class="kw">MakeObj</span>(CGI_hg38, {
  CGI_hg38 &lt;-<span class="st"> </span><span class="kw">ImportCpGIslandExt</span>(organism<span class="op">$</span>UCSC<span class="op">$</span>genome)
})</code></pre></div>
<p>This script generates and saves <code>EGA91_human</code> and <code>CGI_hg38</code> R objects as summarized in the table below.</p>
<table>
<thead>
<tr class="header">
<th>R data object</th>
<th>Organism</th>
<th>Genome build</th>
<th>Source</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>EGA91_human</td>
<td>Homo sapiens</td>
<td>GRCh38 (hg38)</td>
<td>Ensembl genes (rel. 91, Dec. 2017)</td>
</tr>
<tr class="even">
<td>CGI_hg38</td>
<td>Homo sapiens</td>
<td>GRCh38 (hg38)</td>
<td>UCSC / cpgIslandExt.txt.gz</td>
</tr>
</tbody>
</table>
<p> </p>
</div>
<div id="grch37.p13-hg19" class="section level4">
<h4><span class="header-section-number">2.2.2.2</span> GRCh37.p13 / hg19</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># =============================================================================.</span>
<span class="co"># Human genome annotations (Ensembl GRCh37.p13 release 75)</span>
<span class="co"># -----------------------------------------------------------------------------.</span>
organism &lt;-<span class="st"> </span><span class="kw">list</span>(
  <span class="dt">name  =</span> <span class="st">&quot;Homo sapiens&quot;</span>,
  <span class="dt">taxid =</span> <span class="dv">9606</span>,
  <span class="dt">UCSC    =</span> <span class="kw">list</span>(<span class="dt">genome =</span> <span class="st">&quot;hg19&quot;</span>),
  <span class="dt">Ensembl =</span> <span class="kw">list</span>(<span class="dt">release =</span> <span class="dv">75</span>),
  <span class="dt">BSg_pkg =</span> <span class="st">&quot;BSgenome.Hsapiens.UCSC.hg19&quot;</span>,
  <span class="dt">TxDb =</span> <span class="kw">list</span>(
    <span class="dt">source =</span> <span class="st">&quot;Ensembl GRCh37.p13 release 75 (February 2014)&quot;</span>,
    <span class="dt">portal =</span> <span class="st">&quot;http://Feb2014.archive.ensembl.org/Homo_sapiens&quot;</span>,
    <span class="dt">gtf =</span> <span class="kw">paste0</span>(
      <span class="st">&quot;ftp://ftp.ensembl.org/pub/&quot;</span>,
      <span class="st">&quot;release-75/gtf/homo_sapiens/Homo_sapiens.GRCh37.75.gtf.gz&quot;</span>
    )
  ),
  <span class="dt">BioMart =</span> <span class="kw">list</span>(
    <span class="dt">host     =</span> <span class="st">&quot;Feb2014.archive.ensembl.org&quot;</span>, <span class="co"># &quot;grch37.ensembl.org&quot;</span>
    <span class="dt">database =</span> <span class="st">&quot;ensembl&quot;</span>,
    <span class="dt">dataset  =</span> <span class="st">&quot;hsapiens_gene_ensembl&quot;</span>
  )
)

<span class="co"># Gene features from Ensembl</span>
<span class="kw">MakeObj</span>(EGA75_human, {
  EGA75_human &lt;-<span class="st"> </span><span class="kw">BuildGeneFeatures</span>(organism)
})

<span class="co"># CpG Islands from UCSC</span>
<span class="kw">MakeObj</span>(CGI_hg19, {
  CGI_hg19 &lt;-<span class="st"> </span><span class="kw">ImportCpGIslandExt</span>(organism<span class="op">$</span>UCSC<span class="op">$</span>genome)
})</code></pre></div>
<p>This script generates and saves <code>EGA75_human</code> and <code>CGI_hg19</code> R objects as summarized in the table below.</p>
<table>
<thead>
<tr class="header">
<th>R data object</th>
<th>Organism</th>
<th>Genome build</th>
<th>Source</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>EGA75_human</td>
<td>Homo sapiens</td>
<td>GRCh37 (hg19)</td>
<td>Ensembl genes (rel. 75, Feb. 2014)</td>
</tr>
<tr class="even">
<td>CGI_hg19</td>
<td>Homo sapiens</td>
<td>GRCh37 (hg19)</td>
<td>UCSC / cpgIslandExt.txt.gz</td>
</tr>
</tbody>
</table>
<p> </p>
</div>
</div>
<div id="mouse-genome-1" class="section level3">
<h3><span class="header-section-number">2.2.3</span> Mouse genome</h3>
<div id="grcm38.p5-mm10-1" class="section level4">
<h4><span class="header-section-number">2.2.3.1</span> GRCm38.p5 / mm10</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># =============================================================================.</span>
<span class="co"># Mouse genome annotations (Ensembl GRCm38.p5 release 91)</span>
<span class="co"># -----------------------------------------------------------------------------.</span>
organism &lt;-<span class="st"> </span><span class="kw">list</span>(
  <span class="dt">name    =</span> <span class="st">&quot;Mus musculus&quot;</span>,
  <span class="dt">taxid   =</span> <span class="dv">10090</span>,
  <span class="dt">UCSC    =</span> <span class="kw">list</span>(<span class="dt">genome =</span> <span class="st">&quot;mm10&quot;</span>),
  <span class="dt">Ensembl =</span> <span class="kw">list</span>(<span class="dt">release =</span> <span class="dv">91</span>),
  <span class="dt">BSg_pkg =</span> <span class="st">&quot;BSgenome.Mmusculus.UCSC.mm10&quot;</span>,
  <span class="dt">TxDb =</span> <span class="kw">list</span>(
    <span class="dt">source =</span> <span class="st">&quot;Ensembl GRCm38.p5 release 91 (December 2017)&quot;</span>,
    <span class="dt">portal =</span> <span class="st">&quot;http://Dec2017.archive.ensembl.org/Mus_musculus&quot;</span>,
    <span class="dt">gtf =</span> <span class="kw">paste0</span>(
      <span class="st">&quot;ftp://ftp.ensembl.org/pub/&quot;</span>,
      <span class="st">&quot;release-91/gtf/mus_musculus/Mus_musculus.GRCm38.91.gtf.gz&quot;</span>
    )
  ),
  <span class="dt">BioMart =</span> <span class="kw">list</span>(
    <span class="dt">host     =</span> <span class="st">&quot;www.ensembl.org&quot;</span>, <span class="co"># Dec2017.archive.ensembl.org</span>
    <span class="dt">database =</span> <span class="st">&quot;ensembl&quot;</span>,
    <span class="dt">dataset  =</span> <span class="st">&quot;mmusculus_gene_ensembl&quot;</span>
  )
)

<span class="co"># Gene features from Ensembl</span>
<span class="kw">MakeObj</span>(EGA91_mouse, {
  EGA91_mouse &lt;-<span class="st"> </span><span class="kw">BuildGeneFeatures</span>(organism)
})

<span class="co"># CpG Islands from UCSC</span>
<span class="kw">MakeObj</span>(CGI_mm10, {
  CGI_mm10 &lt;-<span class="st"> </span><span class="kw">ImportCpGIslandExt</span>(organism<span class="op">$</span>UCSC<span class="op">$</span>genome)
})</code></pre></div>
<p>This script generates and saves <code>EGA91_mouse</code> and <code>CGI_mm10</code> R objects as summarized in the table below.</p>
<table>
<thead>
<tr class="header">
<th>R data object</th>
<th>Organism</th>
<th>Genome build</th>
<th>Source</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>EGA91_mouse</td>
<td>Mus musculus</td>
<td>GRCm38 (mm10)</td>
<td>Ensembl genes (rel. 91, Dec. 2017)</td>
</tr>
<tr class="even">
<td>CGI_mm10</td>
<td>Mus musculus</td>
<td>GRCm38 (mm10)</td>
<td>UCSC / cpgIslandExt.txt.gz</td>
</tr>
</tbody>
</table>
<p> </p>
</div>
<div id="ncbim37-mm9" class="section level4">
<h4><span class="header-section-number">2.2.3.2</span> NCBIM37 / mm9</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># =============================================================================.</span>
<span class="co"># Mouse genome annotations (Ensembl NCBIM37 release 67)</span>
<span class="co"># -----------------------------------------------------------------------------.</span>
organism &lt;-<span class="st"> </span><span class="kw">list</span>(
  <span class="dt">name    =</span> <span class="st">&quot;Mus musculus&quot;</span>,
  <span class="dt">taxid   =</span> <span class="dv">10090</span>,
  <span class="dt">UCSC    =</span> <span class="kw">list</span>(<span class="dt">genome =</span> <span class="st">&quot;mm9&quot;</span>),
  <span class="dt">Ensembl =</span> <span class="kw">list</span>(<span class="dt">release =</span> <span class="dv">67</span>),
  <span class="dt">BSg_pkg =</span> <span class="st">&quot;BSgenome.Mmusculus.UCSC.mm9&quot;</span>,
  <span class="dt">TxDb =</span> <span class="kw">list</span>(
    <span class="dt">source =</span> <span class="st">&quot;Ensembl NCBIM37 release 67 (May 2012)&quot;</span>,
    <span class="dt">portal =</span> <span class="st">&quot;http://May2012.archive.ensembl.org/Mus_musculus&quot;</span>,
    <span class="dt">gtf =</span> <span class="kw">paste0</span>(
      <span class="st">&quot;ftp://ftp.ensembl.org/pub/&quot;</span>,
      <span class="st">&quot;release-67/gtf/mus_musculus/Mus_musculus.NCBIM37.67.gtf.gz&quot;</span>
    )
  ),
  <span class="dt">BioMart =</span> <span class="kw">list</span>(
    <span class="dt">host     =</span> <span class="st">&quot;May2012.archive.ensembl.org&quot;</span>,
    <span class="dt">database =</span> <span class="st">&quot;ensembl&quot;</span>,
    <span class="dt">dataset  =</span> <span class="st">&quot;mmusculus_gene_ensembl&quot;</span>
  )
)

<span class="co"># Gene features from Ensembl</span>
<span class="kw">MakeObj</span>(EGA67_mouse, {
  EGA67_mouse &lt;-<span class="st"> </span><span class="kw">BuildGeneFeatures</span>(organism)
})

<span class="co"># CpG Islands from UCSC</span>
<span class="kw">MakeObj</span>(CGI_mm9, {
  CGI_mm9 &lt;-<span class="st"> </span><span class="kw">ImportCpGIslandExt</span>(organism<span class="op">$</span>UCSC<span class="op">$</span>genome)
})</code></pre></div>
<p>This script generates and saves <code>EGA67_mouse</code> and <code>CGI_mm9</code> R objects as summarized in the table below.</p>
<table>
<thead>
<tr class="header">
<th>R data object</th>
<th>Organism</th>
<th>Genome build</th>
<th>Source</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>EGA67_mouse</td>
<td>Mus musculus</td>
<td>NCBIM37 (mm9)</td>
<td>Ensembl genes (rel. 67, May 2012)</td>
</tr>
<tr class="even">
<td>CGI_mm9</td>
<td>Mus musculus</td>
<td>NCBIM37 (mm9)</td>
<td>UCSC / cpgIslandExt.txt.gz</td>
</tr>
</tbody>
</table>
<p> </p>
</div>
</div>
</div>
</div>
<div id="genomic-tiling-with-regular-bins" class="section level1">
<h1><span class="header-section-number">3</span> Genomic tiling with regular bins</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ANN &lt;-<span class="st"> </span>EGA91_mouse
CGI &lt;-<span class="st"> </span>CGI_mm10
BLK &lt;-<span class="st"> </span>mm10_blacklist

<span class="co"># Make 4kb windows every 1kb over the whole genome</span>
WGT &lt;-<span class="st"> </span><span class="kw">GenomicTiling</span>(ANN<span class="op">$</span>seqinfo, <span class="dt">s =</span> <span class="dv">1000</span>, <span class="dt">w =</span> <span class="dv">4000</span>)

<span class="co"># Functional annotation of genomic bins</span>
grp &lt;-<span class="st"> </span><span class="kw">factor</span>(
  <span class="kw">rep</span>(<span class="dv">1</span>, <span class="kw">length</span>(WGT)), <span class="dt">levels =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dt">ordered =</span> T,
  <span class="dt">labels =</span> <span class="kw">c</span>(<span class="st">&quot;intergenic&quot;</span>, <span class="st">&quot;genic&quot;</span>, <span class="st">&quot;TSS&quot;</span>, <span class="st">&quot;CGI&quot;</span>, <span class="st">&quot;blacklist&quot;</span>)
)
grp[<span class="kw">which</span>(<span class="kw">countOverlaps</span>(WGT, ANN<span class="op">$</span>GNU) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)] &lt;-<span class="st"> &quot;genic&quot;</span>
grp[<span class="kw">which</span>(<span class="kw">countOverlaps</span>(WGT, ANN<span class="op">$</span>TSS) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)] &lt;-<span class="st"> &quot;TSS&quot;</span>
grp[<span class="kw">which</span>(<span class="kw">countOverlaps</span>(WGT, CGI) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)] &lt;-<span class="st"> &quot;CGI&quot;</span>
grp[<span class="kw">which</span>(<span class="kw">countOverlaps</span>(WGT, BLK) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)] &lt;-<span class="st"> &quot;blacklist&quot;</span>

WGT<span class="op">$</span>category &lt;-<span class="st"> </span>grp</code></pre></div>
</div>
<div id="defining-other-relevant-regions" class="section level1">
<h1><span class="header-section-number">4</span> Defining other relevant regions</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ANN &lt;-<span class="st"> </span>EGA91_mouse
CGI &lt;-<span class="st"> </span>CGI_mm10
BLK &lt;-<span class="st"> </span>mm10_blacklist

<span class="co"># Define promoter regions</span>
TSS &lt;-<span class="st"> </span><span class="kw">resize</span>(ANN<span class="op">$</span>TSS, <span class="dt">width =</span> <span class="dv">4000</span>, <span class="dt">fix =</span> <span class="st">&quot;center&quot;</span>, <span class="dt">use.names =</span> F)

<span class="co"># Define CGI regions as 4kb windows centered at CpG-islands annotated by UCSC</span>
CGI &lt;-<span class="st"> </span><span class="kw">resize</span>(CGI, <span class="dt">width =</span> <span class="dv">4000</span>, <span class="dt">fix =</span> <span class="st">&quot;center&quot;</span>, <span class="dt">use.names =</span> F)
CGI &lt;-<span class="st"> </span><span class="kw">with</span>(ANN, <span class="kw">CleanupGRanges</span>(CGI, seqinfo, <span class="dt">organism =</span> organism<span class="op">$</span>name))</code></pre></div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
